A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+### Earliest Deadline First (EDF) Scheduler
+
+`weighted_round_robin` introduces a scheduler that manages weighted subchannel selection. A new scheduler instance is created every time weights are updated.  A state object is passed when created so that the scheduler can start where it last left across multiple instances. The scheduler must be light and cheap to create and allow concurrent pick calls. The scheduler returns the index in the input weight array and the caller has to map the index to a subchannel.
+
+The scheduler resides in the picker. When compared to `round_robin`, the scheduler conceptually replaces the monotonically increasing index counter of the picker. The picker in `weighted_round_robin` otherwise works the same externally as in `round_robin`.
+
+In C/C++, the scheduler is reference counted and its access is mutex protected. We may consider optimization in the future such as using RCU synchronization when available. Java/Go implementation accesses the scheduler with an atomic reference object.
+
+The picker starts a timer to swap the scheduler to update update weights every `weight_update_period`. This timer iterates over the subchannel list (wrapper) of the picker, loads the latest weight from each, builds a new scheduler and replaces it with the current one. Note this timer does not run in the control plane, but weights are atomic and reading weights does not require extra synchronization with the control plane. When the scheduler is swapped, ongoing pick calls that already took a reference of the outgoing scheduler can continue to refer to the old scheduler until the call returns.
+
+Note only subchannels in `READY` state are passed to the scheduler. Subchannels that are included with a zero weight are still picked but with the average weight. When less than two subchannels have load info, it degrades to `round_robin`.
+
+`weighted_round_robin` uses the earliest deadline first (EDF) scheduling with a monotonically increasing counter as the state. This counter is persisted across multiple scheduler instances for continuity between weight updates.
+
+```
+EDFScheduler {
+  // Creates a scheduler with given weight values.
+  // `next_seq_fn` returns a monotonically increasing
+  // number that wraps, must allow concurrent calls.
+  function EDFScheduler New(float[] weights, lambda () -> uint32 next_seq_fn);
+  // Returns the index of the subchannel selected.
+  // Can be called concurrently.
+  function Pick();
+};
+```
+
+### Control Plane Behaviors
+
+`weighted_round_robin` control plane works identically to `round_robin` except the handling load reports and weights and the scheduler. The current `round_robin` implementation in each language should serve as full reference. This section outlines the differences and key behaviors.
+
+#### Aggregate Connectivity State
```

Comment:
Please combine this with the "Managing Subchannels" section and call it something like "Handling Subchannel Connectivity State Notifications".

Reply:
Done.